### PR TITLE
Realistic benchmark

### PR DESCRIPTION
**Pull request by Raynos**
I "fixed" the benchmark to make it realistic. This means the benchmarks uses the logical retryable requests. It also uses stats and other things that were being skipped.

I also added two more modes to it

 - `benchmarks/take_relay.sh`
 - `benchmarks/take_trace.sh`

These two spawn extra processes to test the relaying and tracing features of hyperbahn.

r: @rf @jcorbin

### CODE REVIEW

**Comment by Raynos** [comment]
Python failed the build.

**Comment by Raynos** [comment]
```
raynos at raynos-MacBookPro  ~/uber/tchannel/node on realistic-benchmark
$ node benchmarks/compare.js benchmarks/master.json benchmarks/realistic-benchmark.json 
GET 16KiB, 1000/0  rate: hi-diff:  -2599.58 (-29.3%)
GET 16KiB, 10000/0 rate: hi-diff:  -2151.94 (-27.5%)
GET 16KiB, 20000/0 rate: hi-diff:  -2562.50 (-32.0%)
GET 4B, 1000/0     rate: hi-diff:  -3454.11 (-30.7%)
GET 4B, 10000/0    rate: hi-diff:  -2937.72 (-27.1%)
GET 4B, 20000/0    rate: hi-diff:  -3578.04 (-29.1%)
GET 4KiB, 1000/0   rate: hi-diff:  -2514.58 (-28.2%)
GET 4KiB, 10000/0  rate: hi-diff:  -2671.74 (-28.2%)
GET 4KiB, 20000/0  rate: hi-diff:  -3260.60 (-31.6%)
PING, 1000/0       rate: hi-diff:  -3844.43 (-32.8%)
PING, 10000/0      rate: hi-diff:  -2543.24 (-24.2%)
PING, 20000/0      rate: hi-diff:  -4180.81 (-33.7%)
SET 16KiB, 1000/0  rate: hi-diff:  -2888.26 (-31.2%)
SET 16KiB, 10000/0 rate: hi-diff:  -1380.51 (-17.2%)
SET 16KiB, 20000/0 rate: hi-diff:  -1600.86 (-18.8%)
SET 4B, 1000/0     rate: hi-diff:  -2954.68 (-27.2%)
SET 4B, 10000/0    rate: hi-diff:  -3295.74 (-29.6%)
SET 4B, 20000/0    rate: hi-diff:  -3566.98 (-29.4%)
SET 4KiB, 1000/0   rate: hi-diff:  -3150.99 (-31.5%)
SET 4KiB, 10000/0  rate: hi-diff:  -3009.71 (-29.0%)
SET 4KiB, 20000/0  rate: hi-diff:  -2569.98 (-23.9%)
```

```
$ node benchmarks/compare.js benchmarks/master.json benchmarks/relay-realistic-benchmark.json 
GET 16KiB, 1000/0  rate: hi-diff:  -5412.45 (-61.1%)
GET 16KiB, 10000/0 rate: hi-diff:  -4817.21 (-61.5%)
GET 16KiB, 20000/0 rate: hi-diff:  -4902.43 (-61.3%)
GET 4B, 1000/0     rate: hi-diff:  -6931.93 (-61.7%)
GET 4B, 10000/0    rate: hi-diff:  -6627.47 (-61.1%)
GET 4B, 20000/0    rate: hi-diff:  -7983.78 (-65.0%)
GET 4KiB, 1000/0   rate: hi-diff:  -5030.12 (-56.4%)
GET 4KiB, 10000/0  rate: hi-diff:  -5840.06 (-61.7%)
GET 4KiB, 20000/0  rate: hi-diff:  -6652.15 (-64.4%)
PING, 1000/0       rate: hi-diff:  -7312.31 (-62.3%)
PING, 10000/0      rate: hi-diff:  -6237.93 (-59.4%)
PING, 20000/0      rate: hi-diff:  -8297.73 (-66.8%)
SET 16KiB, 1000/0  rate: hi-diff:  -5814.21 (-62.7%)
SET 16KiB, 10000/0 rate: hi-diff:  -4634.08 (-57.9%)
SET 16KiB, 20000/0 rate: hi-diff:  -5319.13 (-62.6%)
SET 4B, 1000/0     rate: hi-diff:  -6449.72 (-59.3%)
SET 4B, 10000/0    rate: hi-diff:  -6868.90 (-61.7%)
SET 4B, 20000/0    rate: hi-diff:  -7907.99 (-65.1%)
SET 4KiB, 1000/0   rate: hi-diff:  -6523.11 (-65.2%)
SET 4KiB, 10000/0  rate: hi-diff:  -6626.95 (-63.9%)
SET 4KiB, 20000/0  rate: hi-diff:  -6881.28 (-63.9%)
```

File: node/benchmarks/Makefile [hunk 1]
```diff
@@ -0,0 +1,6 @@
+kill-dead-benchmarks:
+	ps aux | grep nodejs-benchmarks | awk '{ print $$2; }' | xargs kill;
```
Comment (jcorbin) [inline node/benchmarks/Makefile]: pkill -f nodejs-benchmarks

File: node/benchmarks/Makefile [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+
+top-benchmark:
+	top -cp `ps aux | grep nodejs-benchmarks | grep -v grep | awk '{ print $$2; }' | tr "\\n" "," | sed 's/,$$//'`;
```
Comment (jcorbin) [inline node/benchmarks/Makefile]: top -cp $(pgrep -f nodejs--benchmarks)

File: node/benchmarks/Makefile [hunk 3]
```diff
@@ -0,0 +1,6 @@
+kill-dead-benchmarks:
```
Comment (jcorbin) [inline node/benchmarks/Makefile]: Add line at top:
```
.PHONY: kill-dead-benchmarks top-benchmark
```

File: node/benchmarks/index.js [hunk 4]
```diff
@@ -20,21 +20,28 @@
 
 'use strict';
 
+/*eslint no-console: 0*/
```
Comment (jcorbin) [inline node/benchmarks/index.js]: we're not yet using that linter, let's not just start adding habitual exceptions until we do... plus won't it complain about the comment formatting here?

File: node/benchmarks/take_relay.sh [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+set -e
+set -x
+
+rev=$(git rev-parse --short HEAD)
+sym=$(git symbolic-ref HEAD)
+sym=${sym##*/}
+
+node benchmarks/index.js --relay -o benchmarks/relay-$rev.json -- --relay -m 5 -c 10 -p 1000,10000,20000 -s 4,$(( 2 ** 12 )),$(( 2 ** 14 ))
+ln -sf relay-$rev.json benchmarks/relay-$sym.json
```
Comment (jcorbin) [inline node/benchmarks/take_relay.sh]: nice extension of the existing pattern, maybe we should prefix all the normal ones something like "direct-" or "server-" ?

File: node/test/lib/fake-egress-nodes.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    return self.topology[serviceName].indexOf(self.hostPort) >= 0;
+};
+
+FakeEgressNodes.prototype.getRandomNodes =
+function getRandomNodes(serviceName) {
+    var self = this;
+
+    var hosts = [];
+
+    for (var i = 0; i < self.kValue; i++) {
```
Comment (jcorbin) [inline node/test/lib/fake-egress-nodes.js]: Not actually a good random sampler, but doesn't matter #justsayin.

How to do it right: select the top-k randomly chosen scores, e.g. this one-liner in Python:
```
[
    letter
    for score, letter in heapq.nlargest(10, (
        (random.random(), letter)
        for letter in string.ascii_letters))
]
```

I.e.:
- get you a min heap implementation
- iterate your input array, for each item
  - compute a random score
  - if heap is under k in size, push score and item
  - else if score is greater than head of heap, replace and sift down
- strip the scores off and return just the items out of your heap

File: node/test/lib/fake-egress-nodes.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+    var hostPorts = self.topology[serviceName];
+
+    // A random service; pick k random things
+    if (!hostPorts) {
+        hostPorts = self.getRandomNodes(serviceName);
+    }
+
+    var result = Object.create(null);
+    hostPorts.forEach(function buildResult(hostPort, index) {
```
Comment (jcorbin) [inline node/test/lib/fake-egress-nodes.js]: ahh yes, forEach where you use the index argument and a oneline body... such Array#forEach, much slow

**Comment by jcorbin** [comment]
:+1: Great stuff, optional and forward intended feedback.

File: node/benchmarks/index.js (hunk 4 above)
Comment (Raynos) [inline node/benchmarks/index.js]: In preparation of upgrading tchannel to uber-standard I'm writing all new code to be uber-standard compliant. Mostly relying on my text editor plugin.

I can remove the comment though.

File: node/test/lib/fake-egress-nodes.js (hunk 7 above)
Comment (Raynos) [inline node/test/lib/fake-egress-nodes.js]: such fake so tests. But will optimize for benchmarks :)

**Pull request merged**